pkg/ports/publish.go

```diff
@@ -216,23 +216,39 @@ func collectPorts(seen map[int32][]int32, seenHostnames map[string]struct{}, por
 			port.Port = port.TargetPort
 		}
 
-		if targets, ok := seen[port.Port]; ok {
-			// Check for special case: the same port is exposed on multiple hostnames, so keep both.
-			if port.Hostname != "" {
-				for _, t := range targets {
-					if t == port.TargetPort {
-						// Same port and target port but different hostnames, so keep both
-						seen[port.Port] = append(targets, port.TargetPort)
-						seenHostnames[port.Hostname] = struct{}{}
-						result = append(result, port)
+		if seenPortDefs, ok := seen[port.Port]; ok {
+			discard := false
+			for _, p := range seenPortDefs {
+				if port.TargetPort == p.TargetPort {
+					if port.Hostname != "" {
+						// OK: Same port and target port (and potentially protocol) but different hostnames, so keep both
 						break
 					}
+					if !(port.Protocol == v1.ProtocolUDP && p.Protocol != v1.ProtocolUDP ||
+						port.Protocol != v1.ProtocolUDP && p.Protocol == v1.ProtocolUDP) {
```

Comment:
This could be logically cleaned up, I'd also recommend putting the `&` blocks in parens so it's easier to understand how it gets interpreted. It should be:
```
if (port.Protocol != v1.ProtocolUDP && p.Protocol != v1.ProtocolUDP) || (port.Protocol == v1.ProtocolUDP && p.Protocol == v1.ProtocolUDP) {
```

Reply:
To explain this further, you can view `port = udp` as a logical boolean value `A` (thus `port != udp` is `!A`), and `p = udp` as `B` (thus `p != udp` is `!B`). Your initial logical statement is `!(A!B|!AB)`, which simplifies to `!A!B|AB`